builder/codebuild/codebuild.go

```diff
@@ -0,0 +1,350 @@
+	// The repository which is being built
+	Repository string
+
+	// The commit sha at which to do the build
+	Sha string
+
+	// The branch at which the build is happening
+	Branch string
+}
+
+type LogInfo struct {
+	// Name of the cloudwatch group
+	GroupName string
+
+	// Name of cloudwatch stream
+	StreamName string
+}
+
+// NewBuilder returns a new Builder backed by the codebuild client.
+func NewBuilder(config client.ConfigProvider) *Builder {
+	return &Builder{
+		codebuild: codebuild.New(config),
+	}
+}
+
+// NewBuilderFromEnv returns a new Builder with a codebuild client
+// configured from the standard Docker environment variables.
+func NewBuilderFromEnv() (*Builder, error) {	
+
+	serviceRole := os.Getenv("CODEBUILD_SERVICE_ROLE")
+
+	if serviceRole == "" {
+		return nil, errors.New("CODEBUILD_SERVICE_ROLE must be set when using codebuild builder")
+	}
+
+	image := os.Getenv("CODEBUILD_IMAGE")
+
+	if image == "" {
+		image = DefaultCodebuildImage
+	}
+
+	computeType := os.Getenv("CODEBUILD_COMPUTE_TYPE")
+
+	if computeType == "" {
+		computeType = DefaultCodebuildComputeType
+	}
+
+	dockerUsername := os.Getenv("DOCKER_USERNAME")
+	dockerPassword := os.Getenv("DOCKER_PASSWORD")
+
+	if dockerUsername == "" || dockerPassword == "" {
+		return nil, errors.New("DOCKER_USERNAME and DOCKER_PASSWORD env vars must be set when using codebuild builder")
+	}
+
+	sess := session.Must(session.NewSession())
+
+	return &Builder{
+		codebuild: codebuild.New(sess),
+		ServiceRole: serviceRole,
+		Image: image,
+		ComputeType: computeType,
+		DockerUsername: dockerUsername,
+		DockerPassword: dockerPassword,
+	}, nil
+}
+
+
+// Build executes the docker image.
+func (b *Builder) Build(ctx context.Context, w io.Writer, opts builder.BuildOptions) (image string, err error) {
+	image = fmt.Sprintf("%s:%s", opts.Repository, opts.Sha)
+	err = b.build(ctx, w, opts)
+	return
+}
+
+// Build executes the codebuild image.
+func (b *Builder) build(ctx context.Context, w io.Writer, opts builder.BuildOptions) error {
+	
+	projectName := strings.Join([]string{
+		"conveyor",
+		strings.Replace(opts.Repository, "/", "_", -1),
+	}, "-")
+
+	startBuild, err := b.startBuild(opts, projectName)
+
+	if err != nil {
+
+		awsErr, ok := err.(awserr.Error)
+
+    	if ok && awsErr.Code() == "ResourceNotFoundException" {
+
+	        _, err = b.createProject(opts, projectName)
+
+			if err != nil {
+				return err
+			} 
+
+			startBuild, err = b.startBuild(opts, projectName)
+			
+			if err != nil {
+				log.Fatal(err)
+				return err
+			}
+	        
+	    } else {
+	    	log.Fatal(err)
+	    	return err
+	   	}
+
+    }
+
+    logInfo, err := b.getLogInfo(*startBuild.Build.Id)
+
+    if err != nil {
+    	log.Fatal(err)
+    	return err
+    }
+
+
+    sess := session.Must(session.NewSession())
+
+    r, err := cloudwatch.NewLogger(sess, logInfo.GroupName).Open(logInfo.StreamName)
+   	
+   	if err != nil {
+   		log.Fatal(err)
+   		return err
+   	}
+
+    _, err = io.Copy(w, r)
```

Comment:
One thing I wasn't sure about is whether or not this will ever actually return. Looking at https://github.com/ejholmes/cloudwatch/blob/master/reader.go, it never returns an `io.EOF`, which means the reader goroutine continues spinning forever I think.

Did you get a chance to test this?

If it doesn't return, then we may need to poll `BatchGetBuilds` in parallel and close the Reader when the build is done.